### PR TITLE
ControllerEmu: Separate ControlGroup from ControllerEmu

### PR DESCRIPTION
**Pull request by lioncash**
ControllerEmu, the class, is essentially acting like a namespace for ControlGroup. This makes it impossible to forward declare any of the internals. It also globs a bunch of classes together which is kind of a pain to manage.

This splits ControlGroup and the classes it contains into their own source files and situates them all within a namespace, which gets them out of global scope.

Since this allows forward declarations for the once-internal classes, it now requires significantly less files to be rebuilt if anything is changed in the ControllerEmu portion of code.

It does not split out the settings classes from ControlGroup yet, however, as splitting those out will need to have more functional changes made, rather than just essentially moving code from point A to point B.

### CODE REVIEW

File: Source/Core/Core/HW/GCKeyboard.cpp [hunk 1]
```diff
@@ -2,13 +2,16 @@
...
+#include "Core/HW/GCKeyboard.h"
+
 #include <cstring>
 
 #include "Common/Common.h"
 #include "Common/CommonTypes.h"
-#include "Core/HW/GCKeyboard.h"
+
 #include "Core/HW/GCKeyboardEmu.h"
-#include "InputCommon/ControllerInterface/ControllerInterface.h"
```
Comment (ligfx) [inline Source/Core/Core/HW/GCKeyboard.cpp]: Why was this removed? Isn't it still needed to access `g_controller_interface`?

File: Source/Core/Core/HW/GCPadEmu.h [hunk 2]
```diff
@@ -8,7 +8,15 @@
 
 #include "InputCommon/ControllerEmu/ControllerEmu.h"
 
+struct GCPadStatus;
```
Comment (ligfx) [inline Source/Core/Core/HW/GCPadEmu.h]: How does this work? Doesn't it need a full declaration of `GCPadStatus` for the `GetInput()` method?

File: Source/Core/Core/HW/Wiimote.cpp [hunk 3]
```diff
@@ -3,10 +3,15 @@
 // Refer to the license.txt file included.
 
 #include "Core/HW/Wiimote.h"
+
 #include "Common/ChunkFile.h"
+#include "Common/CommonTypes.h"
+
```
Comment (ligfx) [inline Source/Core/Core/HW/Wiimote.cpp]: Why separate all these with newlines (other than Core/HW/Wiimote.h)? clang-format will sort them the same way as you did manually.

File: Source/Core/Core/HW/WiimoteEmu/Speaker.cpp [hunk 4]
```diff
@@ -7,7 +7,9 @@
 #include "AudioCommon/AudioCommon.h"
 #include "Common/CommonTypes.h"
 #include "Common/Logging/Log.h"
+#include "Core/ConfigManager.h"
 #include "Core/HW/WiimoteEmu/WiimoteEmu.h"
+#include "InputCommon/ControllerEmu/ControlGroup/ControlGroup.h"
```
Comment (ligfx) [inline Source/Core/Core/HW/WiimoteEmu/Speaker.cpp]: I don't see where this is used in this file?

File: Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp [hunk 5]
```diff
@@ -10,6 +12,7 @@
 #include "Common/CommonTypes.h"
 #include "Common/MathUtil.h"
 #include "Common/MsgHandler.h"
+
```
Comment (ligfx) [inline Source/Core/Core/HW/WiimoteEmu/WiimoteEmu.cpp]: Same as my other comment, why the newlines?

File: Source/Core/Core/HotkeyManager.cpp [hunk 6]
```diff
@@ -2,11 +2,17 @@
...
 
+#include "Core/HotkeyManager.h"
+
 #include <string>
 #include <vector>
 
 #include "Common/Common.h"
-#include "Core/HotkeyManager.h"
+#include "Common/CommonTypes.h"
+
```
Comment (ligfx) [inline Source/Core/Core/HotkeyManager.cpp]: Same as before, why the newlines?

File: Source/Core/DolphinWX/Debugger/CodeWindow.cpp [hunk 7]
```diff
@@ -28,13 +28,11 @@
 #include "Common/CommonTypes.h"
 #include "Common/StringUtil.h"
 #include "Common/SymbolDB.h"
+#include "Core/ConfigManager.h"
 #include "Core/Core.h"
 #include "Core/Debugger/Debugger_SymbolMap.h"
 #include "Core/Debugger/PPCDebugInterface.h"
 #include "Core/HW/CPU.h"
-#include "Core/HW/Memmap.h"
-#include "Core/HW/SystemTimers.h"
```
Comment (ligfx) [inline Source/Core/DolphinWX/Debugger/CodeWindow.cpp]: Is this related?

File: Source/Core/DolphinWX/Input/InputConfigDiag.cpp [hunk 8]
```diff
@@ -38,21 +40,27 @@
...
 #include "Common/IniFile.h"
 #include "Common/MsgHandler.h"
+
 #include "Core/Core.h"
 #include "Core/HW/GCKeyboard.h"
 #include "Core/HW/GCPad.h"
 #include "Core/HW/Wiimote.h"
 #include "Core/HW/WiimoteEmu/WiimoteEmu.h"
 #include "Core/HotkeyManager.h"
+
```
Comment (ligfx) [inline Source/Core/DolphinWX/Input/InputConfigDiag.cpp]: Why the newlines?

File: Source/Core/Core/HW/Wiimote.cpp (hunk 3 above)
Comment (JosJuice) [inline Source/Core/Core/HW/Wiimote.cpp]: We often put newlines between different categories (Common, Core, DiscIO...) when there are many includes, so that it's easier to read.

File: Source/Core/DolphinWX/Input/InputConfigDiag.h [hunk 9]
```diff
@@ -23,10 +23,10 @@
 #include <wx/spinctrl.h>
 #include <wx/timer.h>
 
-#include "InputCommon/ControlReference/ControlReference.h"
-#include "InputCommon/ControllerEmu/ControllerEmu.h"
+#include "InputCommon/ControllerEmu/ControlGroup/ControlGroup.h"
```
Comment (ligfx) [inline Source/Core/DolphinWX/Input/InputConfigDiag.h]: Can this be forward-declared as well?

File: Source/Core/InputCommon/ControllerEmu/ControlGroup/Extension.cpp [hunk 10]
```diff
@@ -0,0 +1,17 @@
+// Copyright 2017 Dolphin Emulator Project
+// Licensed under GPLv2+
+// Refer to the license.txt file included.
+
+#include "InputCommon/ControllerEmu/ControlGroup/Extension.h"
+
+#include <string>
+
+#include "Common/CommonTypes.h"
+#include "Core/HW/WiimoteEmu/Attachment/Attachment.h"
```
Comment (ligfx) [inline Source/Core/InputCommon/ControllerEmu/ControlGroup/Extension.cpp]: Do this two includes need to be here?

File: Source/Core/Core/HW/GCPadEmu.h (hunk 2 above)
Comment (lioncash) [inline Source/Core/Core/HW/GCPadEmu.h]: It's allowable to use forward declarations on by-value return types and parameters ([see here](http://stackoverflow.com/a/553869) for a list of valid cases)

File: Source/Core/DolphinWX/Debugger/CodeWindow.cpp (hunk 7 above)
Comment (lioncash) [inline Source/Core/DolphinWX/Debugger/CodeWindow.cpp]: I just removed them as they're unnecessary, so it was essentially a "clean-up-as-I-go" sort of thing.

File: Source/Core/DolphinWX/Input/InputConfigDiag.h (hunk 9 above)
Comment (lioncash) [inline Source/Core/DolphinWX/Input/InputConfigDiag.h]: Currently no, because `NumericSetting` is still a nested class within ControlGroup. However this will be changed in a followup.

File: Source/Core/Core/HW/WiimoteEmu/Speaker.cpp (hunk 4 above)
Comment (lioncash) [inline Source/Core/Core/HW/WiimoteEmu/Speaker.cpp]: `m_options` is a `ControlGroup*`

File: Source/Core/Core/HW/GCKeyboard.cpp (hunk 1 above)
Comment (lioncash) [inline Source/Core/Core/HW/GCKeyboard.cpp]: Good catch, I'll re-add that.

File: Source/Core/InputCommon/ControllerEmu/ControlGroup/Extension.cpp (hunk 10 above)
Comment (lioncash) [inline Source/Core/InputCommon/ControllerEmu/ControlGroup/Extension.cpp]: Ah no, they aren't. They can just be replaced with an include to ControllerEmu.h to satisfy the forward decl in the header, good catch.
